src/directives/badge/badge.directive.ts

```diff
@@ -0,0 +1,197 @@
+            this._badgeContent = subject;
+        }
+    }
+    private _badgeContent: string = null;
+
+    /**
+     * Define the badge background color
+     */
+    @Input()
+    get badgeColor(): string {
+        return this._badgeColor.toHex();
+    }
+    set badgeColor(color: string) {
+        this._badgeColor = this.parseThemeColor(color);
+    }
+    private _badgeColor: ThemeColor = this._darkColor;
+
+    /**
+     * Define the badge border color - if unset there is no border
+     */
+    @Input()
+    get badgeBorderColor(): string {
+        return this._badgeBorderColor.toHex();
+    }
+    set badgeBorderColor(color: string) {
+        this._badgeBorderColor = this.parseThemeColor(color);
+    }
+    private _badgeBorderColor: ThemeColor;
+
+    /**
+     * Set the badge vertical position in relation to the parent element
+     */
+    @Input()
+    badgeVerticalPosition: BadgeVerticalPosition = 'above';
+
+    /**
+     * Set the badge horizontal position in relation to the parent element
+     */
+    @Input()
+    badgeHorizontalPosition: BadgeHorizontalPosition = 'after';
+
+    /**
+     * Set if the badge overlaps parent content or flows after parent
+     */
+    @HostBinding('class.ux-badge-overlap')
+    @Input()
+    badgeOverlap: boolean = false;
+
+    /**
+     * Max value that can be displayed - if string measures no. of characters,
+     * if number checks against exact number not no. of characters
+     */
+    @Input() badgeMaxValue: number;
+
+    /**
+     * Badge size (based on CSS styles)
+     */
+    @Input()
+    badgeSize: BadgeSize = 'medium';
+
+    /**
+     * Hide badge from view
+     */
+    @HostBinding('class.ux-badge-hidden')
+    @Input()
+    badgeHidden: boolean = false;
+
+    constructor(
+        private readonly _element: ElementRef<HTMLElement>,
+        private readonly _renderer: Renderer2,
+        private readonly _colorService: ColorService,
+        private readonly _contrastService: ContrastService
+    ) { }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        // get display friendly version of text based on max length and type of val;
+        if (changes.badgeContent || changes.badgeMaxValue) {
+            let finalText: string =
+                (changes.badgeContent && changes.badgeContent.currentValue) || this.badgeContent || null;
+            const maxValue: number =
+                (changes.badgeMaxValue && changes.badgeMaxValue.currentValue) || this.badgeMaxValue || null;
+
+            if (finalText && maxValue && maxValue > 0) {
+                if (this._isNumber && parseInt(finalText) > maxValue) {
+                    finalText = `${maxValue}+`;
+                } else if (finalText.length > maxValue) {
+                    finalText = `${finalText.substr(0, maxValue)}&hellip;`;
+                }
+            }
+
+            this._badgeDisplayContent = finalText;
+        }
+
+        this.updateBadge();
```

Comment:
I think we should improve this slightly. Every time ngOnChanges is called this updateBadge function is called Which re-creates the DOM element every time which could be a bit expensive. We should probably only create the badge element once and then only perform update that are required. So perhaps have individual functions that get called only when required to avoid any unnecessary dom manipulation.